src/main/java/states/DelayPacketState.java

```diff
@@ -0,0 +1,49 @@
+package states;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.InetAddress;
+import java.net.SocketException;
+
+import socket.TFTPDatagramSocket;
+import util.ErrorChecker;
+
+public class DelayPacketState extends ForwardState {
+	public static final int DEFAULT_DELAY = 1000;
+	
+	private ErrorChecker checker;
+	private long delayInMilliseconds;
+	
+	public DelayPacketState(TFTPDatagramSocket socket, InetAddress serverAddress, ErrorChecker checker, long delayInMilliseconds) throws SocketException {
+		super(socket, serverAddress);
+		this.checker = checker;
+		this.delayInMilliseconds = delayInMilliseconds;
+	}
+	
+	public DelayPacketState(TFTPDatagramSocket socket, InetAddress serverAddress) throws SocketException {
+		this(socket, serverAddress, null, DEFAULT_DELAY);
+	}
+	
+	public void setErrorChecker(ErrorChecker checker) {
+		this.checker = checker;
+	}
+	
+	public void setDelay(long delay) {
+		this.delayInMilliseconds = delay;
+	}
+
+	@Override
+	protected void forwardPacket(DatagramPacket packet) throws IOException {
+		if (checker.check(packet)) {
+			LOG.logQuiet("Delaying packet by " + delayInMilliseconds + " ms.");
+			LOG.logVerbose(packet);
+			try {
+				Thread.sleep(delayInMilliseconds);
```

Comment:
Will this work? It's a blocking call so any requests that come in during the sleep will be dropped.

i.e. lose data 2 7000
1. Client does RRQ
2. Error sim forwards
3. Server sends Data 1
4. Error sim forwards
5. Client sends Ack 1
6. Error sim forwards
7. Server sends Data 2
8. Error sim delays Data 2 by 7000 using Thread.sleep()
  - It is not listening on socket.receive() either
9. Client doesn't get Data 2 so sends Ack 1 again, but it gets dropped.

Reply:
^^ Nvm this is not an issue.